libs/remote_storage/src/s3_bucket.rs

```diff
@@ -267,6 +284,9 @@ impl S3Bucket {
         let body = PermitCarrying::new(permit, body);
         let body = TimedDownload::new(started_at, body);
 
+        let cancel_or_timeout = crate::support::cancel_or_timeout(remaining, cancel.clone());
+        let body = crate::support::DownloadStream::new(cancel_or_timeout, body);
+
         Ok(Download {
             metadata,
             etag,
```

Comment:
Can't say any of these variations would look really great to me. I was initially hoping to do a layered approach but with the method parameters it doesn't at least feel natural; basically one would had had just one type handling the concurrency limiting and timing out with the necessary complication.

I guess we are left with `acquire_permit_then_timeout(async { ... })` kind of solution.

Reply:
I moved the cancel to permit acquiring as it used to be in my original PR.